File: DataStructures/ForwardList/forward_list.py
class Node:
    def __init__(self, data, next):
        self.next = next
        self.data = data

class ForwardList:
    def __init__(self):
        self.head = None
        self.sz = 0

    def push_front(self, data):
        self.head = Node(data, self.head)
        self.sz = self.sz + 1

    def pop_front(self):
        if self.head:
            self.head = self.head.next
            self.sz = self.sz - 1

    def size(self):
        return self.sz

    def __getitem__(self, idx):
        tmpNode = self.head
        while idx > 0:
            tmpNode = tmpNode.next
            idx = idx - 1
        return tmpNode.data
    
import unittest

class TestForwardList(unittest.TestCase):
    def test_push_front(self):
        fl = ForwardList()

        for i in range(100):
            fl.push_front(i + 1)

        for i in range(100):
            self.assertEqual(fl[i],  100 - i)

    def test_pop_front(self):
        fl = ForwardList()

        for i in range(100):
            fl.push_front(i + 1)

        for i in range(50):
            fl.pop_front();

        for i in range(0, 50):
            self.assertEqual(fl[i],  50 - i)

        self.assertEqual(fl.size(), 50)

if __name__ == "__main__":
    unittest.main()

File: DataStructures/ForwardList/forward_list.cpp
#include <iostream>
#include <gtest/gtest.h>

template <class T>
struct Node {
    T data;
    Node<T>* next;

    Node(T data, Node<T>* node) {
        this->data = data;
        this->next = node;
    }
};

template <class T>
class ForwardList {
public:
    ForwardList() {
        head = nullptr;
        sz = 0;
    }

    void push_front(T data) {
        head = new Node<T>(data, head);
        // Node<T>* newNode = new Node(data, nullptr);
        // newNode->next = head;
        // head = newNode;
        ++sz;
    }

    void pop_front() {
        Node<T>* nextHead = head->next;
        delete head;
        head = nextHead;
        --sz;
    }

    T& operator [] (size_t index) {
        Node<T>* n = head;
        for (size_t i = 0; i < index; ++i) {
            n = n->next;
        }
        return n->data;
    }

    T& front() {
        if (empty()) {
            throw "front() called on empty forward_list";
        }
        return head->data;
    }

    size_t size() const {
        return sz;
    }

    bool empty() const {
        return sz == 0;
    }

    ~ForwardList() {
        while (!empty())
            pop_front();
    }

private:
    Node<T>* head;
    size_t sz;
};

TEST(PushFrontElements, Subtest_1) {
    ForwardList<int> fl;

    for (int i = 0; i < 100; i++) {
        fl.push_front(i + 1);
    }
    
    for (int i = 0; i < 100; i++) {
        ASSERT_TRUE(fl[i] == 100 - i);
    }
    
    ASSERT_TRUE(fl.size() == 100);
}

TEST(PopFrontElements, Subtest_2) {
    ForwardList<int> fl;

    for (int i = 0; i < 100; i++) {
        fl.push_front(i + 1);
    }

    for (int i = 0; i < 50; i++) {
        fl.pop_front();
    }
    
    for (int i = 0; i < 50; i++) {
        ASSERT_TRUE(fl[i] == 50 - i);
    }
    
    ASSERT_TRUE(fl.size() == 50);
}

int main(int argc, char **argv)
{
  ::testing::InitGoogleTest(&argc, argv);

  return RUN_ALL_TESTS();
}